avalon/tools/workfiles/app.py

```diff
@@ -3,83 +3,123 @@
+                "label": api.Session["AVALON_TASK"]
+            },
+            "version": 1,
+            "user": getpass.getuser(),
+            "comment": ""
+        }
 
-        self.version_spinbox.valueChanged.connect(
+        # Define work files template
+        templates = self.data["project"]["config"]["template"]
+        template = templates.get("workfile",
+                                 "{task[name]}_v{version:0>4}<_{comment}>")
+        self.template = template
+
+        # Build version
+        version = QtWidgets.QWidget()
+        version_spinbox = QtWidgets.QSpinBox()
+        version_spinbox.setMinimum(1)
+        version_spinbox.setMaximum(9999)
+        version_checkbox = QtWidgets.QCheckBox("Next Available Version")
+        version_checkbox.setCheckState(QtCore.Qt.CheckState(2))
+        layout = QtWidgets.QHBoxLayout(version)
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.addWidget(version_spinbox)
+        layout.addWidget(version_checkbox)
+
+        # Build comment
+        comment = QtWidgets.QLineEdit()
+
+        # Build preview
+        preview = QtWidgets.QLabel("Preview filename")
+
+        # Build buttons
+        buttons = QtWidgets.QWidget()
+        layout = QtWidgets.QHBoxLayout(buttons)
+        ok_button = QtWidgets.QPushButton("Ok")
+        cancel_button = QtWidgets.QPushButton("Cancel")
+        layout.addWidget(ok_button)
+        layout.addWidget(cancel_button)
+
+        # Build inputs
+        inputs = QtWidgets.QWidget()
+        layout = QtWidgets.QFormLayout(inputs)
+        layout.addRow("Version:", version)
+        layout.addRow("Comment:", comment)
+        layout.addRow("Preview:", preview)
+
+        # Build layout
+        layout = QtWidgets.QVBoxLayout(self)
+        layout.addWidget(inputs)
+        layout.addWidget(buttons)
+
+        version_spinbox.valueChanged.connect(
             self.on_version_spinbox_changed
         )
-        self.version_checkbox.stateChanged.connect(
+        version_checkbox.stateChanged.connect(
             self.on_version_checkbox_changed
         )
-        self.comment_lineedit.textChanged.connect(self.on_comment_changed)
-        self.ok_button.pressed.connect(self.on_ok_pressed)
-        self.cancel_button.pressed.connect(self.on_cancel_pressed)
+        comment.textChanged.connect(self.on_comment_changed)
+        ok_button.pressed.connect(self.on_ok_pressed)
+        cancel_button.pressed.connect(self.on_cancel_pressed)
 
         # Allow "Enter" key to accept the save.
-        self.ok_button.setDefault(True)
+        ok_button.setDefault(True)
 
         # Force default focus to comment, some hosts didn't automatically
         # apply focus to this line edit (e.g. Houdini)
-        self.comment_lineedit.setFocus()
+        comment.setFocus()
+
+        self.widgets = {
+            "preview": preview,
+            "comment": comment,
+            "version": version,
```

Comment:
Not sure about this one, the `self.widgets` here is collecting widgets from defined vars, but in `ContextBreadcrumb` and the `Window`, the `self.widgets` is init classes first, then making changes.

I wonder if this should be unified, too ?

I ***imaging*** that I may firstly look up the `self.widgets` to have a quick overview on this class, so if there were all just variables would not be very helpful. Again, just imaging :P

Reply:
Ah, yes. I see how they differ. Admittedly I have no preference as I have to get used to either. So I guess this is a good one to take a pick and see what @mottosso brings to the table about how he intended it.

The convention that is [laid out currently seems to first declare solely the widgets at the top](https://github.com/getavalon/core/blob/e40eac945bdeae91dedc38418863624db1bcf613/avalon/tools/convention/view.py#L26) so I guess we should move into that direction.